core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -144,6 +152,40 @@ class LogCleaner(val config: CleanerConfig,
     cleaners.foreach(_.shutdown())
   }
 
+  override def configure(configs: util.Map[String, _]): Unit = {}
+
+  override def reconfigurableConfigs(): util.Set[String] = {
+    LogManager.LogCleanerReconfigurableConfigs.asJava
+  }
+
+  override def validateReconfiguration(configs: util.Map[String, _]): Boolean = {
+    val newConfig = logCleanerConfig(configs)
+    val numThreads = newConfig.numThreads
+    numThreads >= 1 && numThreads >= config.numThreads / 2 && numThreads <= config.numThreads * 2
+  }
+
+  /**
+    * Reconfigure log clean config. This simply stops current log cleaners and creates new ones.
+    * That ensures that if any of the cleaners had failed, new cleaners are created to match the new config.
+    */
+  override def reconfigure(configs: util.Map[String, _]): Unit = {
+    config = logCleanerConfig(configs)
+    shutdown()
+    startup()
+  }
+
+  private def logCleanerConfig(configs: util.Map[String, _]): CleanerConfig = {
+    CleanerConfig(numThreads = configs.get(KafkaConfig.LogCleanerThreadsProp).asInstanceOf[Int],
```

Comment:
It's kind of a pity that `Reconfigurable` has to work with the `Map` interface since it means all this nasty casting. Ideally we could just pass through the `KafkaConfig` object itself and use the nice fields. I am tempted to suggest that we should have a separate `BrokerReconfigurable` interface which allows us to use the nicer object. What do you think?

Reply:
I had thought initially that we should use the same interface internally as we expose externally. But when I came to the final PR for listener update, I ended up doing that with KafkaConfig rather than maps since it was too messy otherwise. So it makes sense to use the same interface here. Updated.